cpp/src/arrow/array.cc

```diff
@@ -283,17 +283,49 @@ std::shared_ptr<Array> StringArray::Slice(int64_t offset, int64_t length) const
 FixedSizeBinaryArray::FixedSizeBinaryArray(const std::shared_ptr<DataType>& type,
     int64_t length, const std::shared_ptr<Buffer>& data,
     const std::shared_ptr<Buffer>& null_bitmap, int64_t null_count, int64_t offset)
-    : PrimitiveArray(type, length, data, null_bitmap, null_count, offset) {
-  DCHECK(type->type == Type::FIXED_SIZE_BINARY);
-  byte_width_ = static_cast<const FixedSizeBinaryType&>(*type).byte_width();
-}
+    : PrimitiveArray(type, length, data, null_bitmap, null_count, offset),
+      byte_width_(static_cast<const FixedSizeBinaryType&>(*type).byte_width()) {}
 
 std::shared_ptr<Array> FixedSizeBinaryArray::Slice(int64_t offset, int64_t length) const {
   ConformSliceParams(offset_, length_, &offset, &length);
   return std::make_shared<FixedSizeBinaryArray>(
       type_, length, data_, null_bitmap_, kUnknownNullCount, offset);
 }
 
+const uint8_t* FixedSizeBinaryArray::GetValue(int64_t i) const {
+  return raw_data_ + (i + offset_) * byte_width_;
+}
+
+// ----------------------------------------------------------------------
+// Decimal
+
+template <>
+Decimal32 DecimalArray::Value(int64_t i) const {
+  Decimal32 result;
+  FromBytes(GetValue(i), &result);
+  return result;
+}
+
+template <>
+Decimal64 DecimalArray::Value(int64_t i) const {
+  Decimal64 result;
+  FromBytes(GetValue(i), &result);
+  return result;
+}
+
+template <>
+Decimal128 DecimalArray::Value(int64_t i) const {
+  Decimal128 result;
+  FromBytes(GetValue(i), IsNegative(i), &result);
```

Comment:
`FromBytes` is now `DCHECK`d and does not return `Status`. If it fails, the something has gone horribly wrong.